### ISSUE TITLE
Directory caching behavior

### ISSUE THREAD

**Issue by aluzzardi** [dagger/dagger#3705]
I noticed that using `Directory()`, changes outside the specified directory will cache bust:

```go
	src := c.Host().Workdir().Directory("sdk/python")
```

However, creating doing a `WithDirectory` won't:

```go
	src := c.Directory().WithDirectory("/", c.Host().Workdir().Directory("sdk/python"))
```

I guess this is normal? This should at least be documented, or perhaps `Directory.Directory()` should have a `copy` flag?

/cc @vito @sipsma

**Comment by plaisted** [comment]
I've been looking for documentation on how different types / layers are cached / invalidated so will second this. Maybe it's more obvious to people who understand buildkit underneath, but it has been mostly trial and error to figure it out for me.

**Comment by aluzzardi** [comment]
@plaisted Right -- we're definitely missing documentation on this.

For the MOST part, it's a "simple" layer cache:

`[A].[B].[C]` --> changes to `B` will invalidate both `B` and `C`, bot not `A`

For example:

```go
c.Container().
  From("A"). // this is `A`
  WithFile("/B", someFile). // this is `B`
  Exec(dagger.ContainerExecOpts{Args: ["C"]}) // this is `C`
```

- Changes to `A` (e.g. the image `A` is different), will invalidate `A` `B` and `C`
- Changes to `B` (e.g. `someFIle` is different than the previous run) will result in `Exec("C")` to be re-executed, but `A` will stay the same

Where it gets more complex is this is recursive. Let's say `someFile` is defined as:

```go
someFile := c.Container().From("alpine:latest").File("/etc/alpine-release")
c.Container().
  From("A"). // this is `A`
  WithFile("/B", someFile). // this is `B`
  Exec(dagger.ContainerExecOpts{Args: ["C"]}) // this is `C`
```

If `alpine:latest` changes (let's say there's a new version), then:

- `From("alpine:latest")` is re-computed
- `.File("/etc/alpine-release")` is re-computed (because `alpine:latest` changed)
- The contents of `.File("/etc/alpine-release")` are different (because it's a different version of alpine)
- This in turn results in a different `someFile`
- Which invalidates the `Exec("C")` cache

However, there's some edge cases (e.g. this issue) which we definitely need to document or get rid of.

/cc @vikram-dagger @sipsma @kpenfound

**Comment by kpenfound** [comment]
Based on a conversation in #help yesterday, I think this is true:
```go
src := c.Host().Workdir().Directory("sdk/python")
```
This will send all of workdir to buildkit and changes outside of sdk/python will bust cache

```go
src := c.Host().Directory("sdk/python")
```
this will just send sdk/python to buildkit and changes outside of sdk/python will not bust cache

We should make this more clear :)

**Comment by plaisted** [comment]
Yep, I follow that in general. For me I'm looking a level down, using the same example what if we grab `/etc/resolv.conf` that may not change between alpine releases. Would a change in files timestamps due to a newer container image trigger a rebuild even if the contents match exactly?

Also what about different call chains, assuming `example.txt` is identical content in both containers `alpine:latest` and `not-alpine:latest`:
```
someFile := c.Container().From("alpine:latest").File("/example.txt")
someFile2 := c.Container().From("not-alpine:latest").File("/example.txt")

c.Container().
  From("A"). // this is `A`
  WithFile("/B", someFile). // this is `B`
  Exec(dagger.ContainerExecOpts{Args: ["C"]}) // this is `C

c.Container().
  From("A"). // this is `AA`
  WithFile("/B", someFile2). // this is `BB`
  Exec(dagger.ContainerExecOpts{Args: ["C"]}) // this is `CC

```

Would A = AA and B = BB and C = CC from a caching perspective since they are identical file wise?
